CORE-6824 Schema Registry/Proto: add 5 missing checks

pgellert opened the pull request:
This implements 5 new compatibility checks for protobuf:
* ONEOF_FIELD_REMOVED
* MULTIPLE_FIELDS_MOVED_TO_ONEOF
* REQUIRED_FIELD_{ADDED,REMOVED}
* FIELD_NAMED_TYPE_CHANGED
* MESSAGE_REMOVED

This gap was discovered and initially implemented on @oleiman's draft PR (https://github.com/redpanda-data/redpanda/pull/18332) aiming to add support for verbose logging. This PR extracts the changes of adding these new compatibility checks to make the review easier. Some unit tests are also added.

Fixes https://redpandadata.atlassian.net/browse/CORE-6824

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v24.2.x
- [x] v24.1.x
- [x] v23.3.x

## Release Notes

### Improvements
* Schema Registry: 5 new compatibility checks are added for protobuf (ONEOF_FIELD_REMOVED, MULTIPLE_FIELDS_MOVED_TO_ONEOF, REQUIRED_FIELD_{ADDED,REMOVED}, FIELD_NAMED_TYPE_CHANGED, MESSAGE_REMOVED)

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

oleiman commented on src/v/pandaproxy/schema_registry/protobuf.cc:
```diff
@@ -499,14 +499,33 @@ struct compatibility_checker {
 
+        // check writer fields
         for (int i = 0; i < writer->field_count(); ++i) {
+            // TODO(oren): Need special handling here?
             if (reader->IsReservedNumber(i) || writer->IsReservedNumber(i)) {
```
fwiw, I don't think I ever got to understanding what this means in the context of compatibility checking. may not be so relevant here (nothing has changed) but maybe Ben can comment.

oleiman replied:
>If the field is required by the reader, but reserved by the writer, it should be incompatible.

That sounds right...maybe the converse as well? I didn't experiment w/ this at all, so maybe a quick test against the reference impl would be prudent.

At any rate this block does need some special treatment then, right? Rather than skipping over reserved fields, it seems like reserved-ness should be part of the predicate below.

pgellert replied:
I don't think `reserved`ness should be part of the compat checks. `reserved` is just an annotation in protobuf, it just ensures that another field cannot be defined with the same tag number / name to help ensure that old fields are not reintroduced.

> If the field is required by the reader, but reserved by the writer, it should be incompatible.

This is true because "reserved by the writer" means that the writer doesn't have the tag and so it is just like introducing a required field which is incompatible.

oleiman replied:
oh yeah, even better. point being that we shouldn't be skipping checks on that basis. 👍

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/52669#019132c3-608f-40b9-b13d-c0005b50a96a
<!--e74e7d36-8714-4587-b11a-04db52484642-->

oleiman commented on src/v/pandaproxy/schema_registry/test/compatibility_protobuf.cc:
```diff
@@ -692,3 +692,15 @@ SEASTAR_THREAD_TEST_CASE(
+      pps::compatibility_level::forward,
+      R"(syntax = "proto3"; message Simple { int32 id = 1; })",
+      R"(syntax = "proto3"; message Simple { oneof wrapper { int32 id = 1; } })"));
+    BOOST_REQUIRE(!check_compatible(
+      pps::compatibility_level::forward,
+      R"(syntax = "proto3"; message Simple { int32 id = 1; int32 new_id = 2; })",
+      R"(syntax = "proto3"; message Simple { oneof wrapper { int32 id = 1; int32 new_id = 2; } })"));
```
this is slightly surprising after reading my note in `protobuf.cc:L544-545` (I don't exactly remember what led me to that conclusion). Does the check also fail in the `backward` direction?

oleiman replied:
oh, and if you can't construct a validating example, you may as well remove the comment 😉

pgellert replied:
I removed this comment now, but I also didn't get what it meant. To me, that comment only makes sense in the other direction, ie. in this commit: https://github.com/redpanda-data/redpanda/pull/22798/commits/d2e710aad27ee6cef4d001b22d880198357ca803

> Does the check also fail in the backward direction?

No it doesn't. I've added a test for this now: https://github.com/redpanda-data/redpanda/pull/22798/commits/9c1670a8ed98b214a882fa79d94842fad9bb32b2

oleiman replied:
>that comment only makes sense in the other direction, ie. in this commit: https://github.com/redpanda-data/redpanda/commit/d2e710aad27ee6cef4d001b22d880198357ca803

Yup, that's the one

oleiman reviewed (commented):
looks good to me. leaving the approval status open since I wrote some of the code.

aanthony-rp reviewed (approved):
nothing standing out but I lack a lot of depth for context here.  looks like oren and ben are digging in.

oleiman commented on src/v/pandaproxy/schema_registry/protobuf.cc:
```diff
@@ -499,14 +499,33 @@ struct compatibility_checker {
-            // A reader may ignore a writer field
-            if (r && !check_compatible(r, writer->field(i))) {
+            // A reader may ignore a writer field iff it is not `required`
+            if (!r && w->is_required()) {
+                return false;
+            } else if (r && !check_compatible(r, w)) {
+                return false;
+            }
```
nitpick: maybe

```suggestion
            return (!r && w->is_required()) || (r && !check_compatible(r, w));
```

Since we're short circuiting the whole function. Though maybe that's even harder to read!

pgellert replied:
Yeah I think the existing one is easier to reason about. (I think this one is also missing a negation from the front to be equivalent with the existing code.)

pgellert commented on src/v/pandaproxy/schema_registry/test/compatibility_protobuf.cc:
```diff
@@ -677,3 +677,10 @@ SEASTAR_THREAD_TEST_CASE(test_protobuf_compatibility_message_removed) {
+
+SEASTAR_THREAD_TEST_CASE(test_protobuf_compatibility_field_name_type_changed) {
+    BOOST_REQUIRE(!check_compatible(
+      pps::compatibility_level::full_transitive,
```
> Please narrow to backward, but I'm not sure this is necessarily incompatible, just notable?
Yeah, it's overly restrictive in this specific case

pgellert commented:
[Force-push](https://github.com/redpanda-data/redpanda/compare/cfd74f74636fe31b6752d64334088c03bdb302f0..6dba17a04d8b7e9a1f1f6481042a8e7a0ffa4e82): 
* Removed `reserved` field checking logic
* Test changes: added new tests, made the compatibility level specific for the added tests, structured all tests as a backward compatibility checks where the first line = new = reader schema and the second line = old = writer schema.

pgellert commented:
[Force-push](https://github.com/redpanda-data/redpanda/compare/cfd74f74636fe31b6752d64334088c03bdb302f0..d2e710aad27ee6cef4d001b22d880198357ca803) (twice): 
* Removed `reserved` field checking logic
* Test changes: added new tests, made the compatibility level specific for the added tests, structured all tests as a backward compatibility checks where the first line = new = reader schema and the second line = old = writer schema.

pgellert commented on src/v/pandaproxy/schema_registry/protobuf.cc:
```diff
@@ -499,14 +499,33 @@ struct compatibility_checker {
+        // check reader required fields
+        for (int i = 0; i < reader->field_count(); ++i) {
+            // TODO(oren): Need special handling here?
+            if (writer->IsReservedNumber(i) || reader->IsReservedNumber(i)) {
```
Done. See: https://github.com/redpanda-data/redpanda/pull/22798#discussion_r1713541319

BenPope reviewed (commented):
Very nice

vbotbuildovich commented:
/backport v24.2.x

vbotbuildovich commented:
/backport v24.1.x

vbotbuildovich commented:
/backport v23.3.x
